sonarlint-vcs/src/main/java/org/sonarsource/sonarlint/core/vcs/GitUtils.java

```diff
@@ -0,0 +1,136 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.vcs;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.errors.IncorrectObjectTypeException;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+
+
+public class GitUtils {
+
+  private static final String BRANCH_REF_PREFIX = "refs/heads/";
+
+  private GitUtils() {
+    // util class
+  }
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  public static List<String> getCommitNamesForRef(String branchName, Git git) {
+    var commitNames = new ArrayList<String>();
+    try {
+      var branchObjectId = git.getRepository().resolve(branchName);
+      if (branchObjectId == null) {
+        return commitNames;
+      }
+      var commits = git.log().add(branchObjectId).call();
+      for (RevCommit commit : commits) {
+        commitNames.add(commit.getName());
+      }
+    } catch (IncorrectObjectTypeException e) {
+      // do nothing, jgit will throw it if branchName can not be resolved to any commit
+    } catch (GitAPIException | IOException e) {
+      LOG.error("Couldn't fetch commits for branch " + branchName, e);
+    }
+    return commitNames;
+  }
+
+  @CheckForNull
+  public static Git getGitForDir(URI projectDirUri) {
+    try {
+      var builder = new FileRepositoryBuilder();
+      var gitUri = new URI(projectDirUri + "/.git");
+      var repository = builder.setGitDir(new File(gitUri)).setMustExist(true).build();
+      return new Git(repository);
+    } catch (IOException | URISyntaxException e) {
+      LOG.error("Couldn't access repository for path " + projectDirUri.getPath());
+    }
+    return null;
+  }
+
+
+  public static Optional<String> electSQBranchForLocalBranch(String branchName, Git git, Set<ServerBranch> serverCandidates) {
```

Comment:
Since the have only one branch that could be a main branch, maybe it would be simpler to pass 2 arguments:
* `Set<String> serverBranchCandidates`
* `String serverMainBranch`
This way you could remove the `ServerBranch` pojo